cli/silence_query.go

```diff
@@ -0,0 +1,143 @@
+package cli
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/prometheus/alertmanager/cli/format"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+var queryFlags *flag.FlagSet
+var queryCmd = &cobra.Command{
+	Use:   "query",
+	Short: "Query silences",
+	Long: `Query Alertmanager silences.
+
+  Amtool has a simplified prometheus query syntax, but contains robust support for
+  bash variable expansions. The non-option section of arguments constructs a list
+  of "Matcher Groups" that will be used to filter your query. The following
+  examples will attempt to show this behaviour in action:
+
+  amtool silence query alertname=foo node=bar
+
+  	This query will match all silences with the alertname=foo and node=bar label
+  	value pairs set.
+
+  amtool silence query foo node=bar
+
+  	If alertname is ommited and the first argument does not contain a '=' or a
+  	'=~' then it will be assumed to be the value of the alertname pair.
+
+  amtool silence query 'alertname=~foo.*'
+
+  	As well as direct equality, regex matching is also supported. The '=~' syntax
+  	(similar to prometheus) is used to represent a regex match. Regex matching
+  	can be used in combination with a direct match.
+
+  amtool silence query alertname=foo node={bar,baz}
+
+  	This query will match all silences with the alertname=foo label value pair
+  	and EITHER node=bar or node=baz.
+
+  amtool silence query alertname=foo{a,b} node={bar,baz}
+
+	Similar to the previous example this query will match all silences with any
+	combination of alertname=fooa or alertname=foob AND node=bar or node=baz.
+				`,
+	RunE: query,
+}
+
+func init() {
+	queryCmd.Flags().Bool("expired", false, "Show expired silences as well as active")
+	queryFlags = queryCmd.Flags()
+}
+
+func fetchSilences(filter *[]labels.Matcher) ([]types.Silence, error) {
+	silenceResponse := alertmanagerSilenceResponse{}
+
+	u, err := GetAlertmanagerURL()
+	if err != nil {
+		return []types.Silence{}, err
+	}
+
+	u.Path = path.Join(u.Path, "/api/v1/silences")
+	if filter != nil {
+		u.RawQuery = url.QueryEscape(MatchersToString(*filter))
+	}
+
+	res, err := http.Get(u.String())
+	if err != nil {
+		return []types.Silence{}, err
+	}
+
+	defer res.Body.Close()
+	decoder := json.NewDecoder(res.Body)
+
+	err = decoder.Decode(&silenceResponse)
+	if err != nil {
+		return []types.Silence{}, err
+	}
+
+	return silenceResponse.Data, nil
+}
+
+func query(cmd *cobra.Command, args []string) error {
```

Comment:
Given the filtering available at the silences endpoint, the merge code and multiple requests can be replaced by a single command line query:

```
amtool silence query alertname=~foo(a|b) node=~(bar|baz)
```

And then this can be sent to the server once parsed as:

```
http://alertmanager.cloudflare.example.biz/api/v1/silences?filter=%7Balertname%3D~%22foo(a%7Cb)%22%2C%20node%3D~%22(bar%7Cbaz)%22%7D
```

I think with this and my other comment the whole merge concept could be removed, which would reduce the number of web requests made to the backend and simplify the code being run in the tool.